File: backend/src/services/badge.service.js
/**
 * Phase 3.3.3: Badge Service
 * Handles badge awarding logic and badge checking
 */

import Badge from '../models/Badge.js';
import BadgeHistory from '../models/BadgeHistory.js';
import User from '../models/User.js';
import Module from '../models/Module.js';
import GameScore from '../models/GameScore.js';
import QuizResult from '../models/QuizResult.js';
import DrillLog from '../models/DrillLog.js';
import logger from '../config/logger.js';

/**
 * Check and award badges for a user
 * @param {string} userId - User ID
 * @param {string} triggerType - Type of action that triggered check ('module', 'game', 'drill', 'streak', 'score')
 * @param {Object} triggerData - Data related to the trigger
 * @returns {Promise<Array>} Array of newly awarded badge IDs
 */
export const checkAndAwardBadges = async (userId, triggerType = null, triggerData = {}) => {
  try {
    const user = await User.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    // Get all active badges
    const badges = await Badge.find({ isActive: true }).lean();

    const newlyAwarded = [];

    for (const badge of badges) {
      // Skip if user already has this badge
      if (user.progress?.badges?.includes(badge.id)) {
        continue;
      }

      // Check if badge matches user's grade level
      if (!badge.gradeLevel.includes('all') && user.grade && !badge.gradeLevel.includes(user.grade)) {
        continue;
      }

      // Check badge criteria
      const meetsCriteria = await checkBadgeCriteria(userId, badge, triggerType, triggerData);

      if (meetsCriteria) {
        // Award badge
        await awardBadge(userId, badge.id, triggerType, triggerData);
        newlyAwarded.push(badge.id);
        logger.info(`Badge awarded: ${badge.id} to user ${userId}`);
      }
    }

    return newlyAwarded;
  } catch (error) {
    logger.error('Check and award badges error:', error);
    return [];
  }
};

/**
 * Check if user meets badge criteria
 */
const checkBadgeCriteria = async (userId, badge, triggerType, triggerData) => {
  const { type, value, moduleCategory, gameType } = badge.criteria;

  try {
    switch (type) {
      case 'module_complete':
        // Badge for completing a specific module or module category
        const user = await User.findById(userId).select('progress');
        if (moduleCategory) {
          // Check if all modules in category are completed
          const modules = await Module.find({ 
            category: moduleCategory, 
            isActive: true 
          }).lean();
          
          const completedModules = user.progress?.completedModules || [];
          const allCompleted = modules.every(module => 
            completedModules.some(id => id.toString() === module._id.toString())
          );
          
          return allCompleted && modules.length > 0;
        } else if (value) {
          // Check if specific module is completed
          const completedModules = user.progress?.completedModules || [];
          return completedModules.some(id => id.toString() === value.toString());
        }
        break;

      case 'module_all':
        // Badge for completing all modules
        const totalModules = await Module.countDocuments({ isActive: true });
        const userModules = await User.findById(userId).select('progress');
        const completedCount = userModules.progress?.completedModules?.length || 0;
        return completedCount >= totalModules && totalModules > 0;

      case 'game_wins':
        // Badge for winning games
        const winCount = parseInt(value) || 1;
        const gameScores = await GameScore.find({ 
          userId,
          gameType: gameType === 'all' ? { $exists: true } : gameType
        });
        
        const wins = gameScores.filter(score => 
          score.maxScore > 0 && score.score >= score.maxScore
        ).length;
        
        return wins >= winCount;

      case 'game_perfect':
        // Badge for perfect game scores
        const perfectGames = await GameScore.find({
          userId,
          gameType: gameType === 'all' ? { $exists: true } : gameType,
          maxScore: { $gt: 0 }
        });
        
        const perfectCount = perfectGames.filter(score => 
          score.score >= score.maxScore
        ).length;
        
        return perfectCount >= (parseInt(value) || 1);

      case 'drill_ack':
        // Badge for acknowledging drills
        const ackCount = parseInt(value) || 1;
        const drillLogs = await DrillLog.countDocuments({ userId });
        return drillLogs >= ackCount;

      case 'drill_speed':
        // Badge for fast drill acknowledgment
        // This would need drill acknowledgment time tracking
        // For now, simplified check
        return false; // Placeholder

      case 'streak_days':
        // Badge for login streaks
        const streakDays = parseInt(value) || 30;
        const userStreak = await User.findById(userId).select('progress.loginStreak');
        return (userStreak.progress?.loginStreak || 0) >= streakDays;

      case 'score_threshold':
        // Badge for reaching preparedness score threshold
        const threshold = parseInt(value) || 80;
        const userScore = await User.findById(userId).select('progress.preparednessScore');
        return (userScore.progress?.preparednessScore || 0) >= threshold;

      default:
        return false;
    }
  } catch (error) {
    logger.error(`Error checking badge criteria for ${badge.id}:`, error);
    return false;
  }

  return false;
};

/**
 * Award badge to user
 * Phase 3.3.3: Enhanced with history tracking
 */
export const awardBadge = async (userId, badgeId, triggerType = 'manual', triggerData = {}) => {
  try {
    const user = await User.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    const badge = await Badge.findOne({ id: badgeId });
    if (!badge) {
      throw new Error('Badge not found');
    }

    // Check if badge was already awarded (check both user progress and history)
    const existingHistory = await BadgeHistory.findOne({ userId, badgeId });
    if (existingHistory) {
      logger.info(`Badge ${badgeId} already awarded to user ${userId}`);
      return false; // Already awarded
    }

    // Add badge if not already present
    if (!user.progress) {
      user.progress = {};
    }
    if (!user.progress.badges) {
      user.progress.badges = [];
    }

    if (!user.progress.badges.includes(badgeId)) {
      user.progress.badges.push(badgeId);
      await user.save();
    }

    // Create badge history record
    await BadgeHistory.create({
      userId,
      badgeId,
      xpEarned: badge.xpReward || 0,
      triggerType,
      triggerData
    });

    // Add XP reward if configured
    // This could trigger score recalculation
    if (badge.xpReward > 0) {
      // XP can be added to user's total XP or used in score calculation
      // Implementation depends on XP system design
    }

    return true;
  } catch (error) {
    // Handle duplicate key error (unique index)
    if (error.code === 11000) {
      logger.info(`Badge ${badgeId} already awarded to user ${userId} (duplicate prevented)`);
      return false;
    }
    logger.error('Award badge error:', error);
    throw error;
  }
};

/**
 * Get all badges for a user
 */
export const getUserBadges = async (userId) => {
  try {
    const user = await User.findById(userId).select('progress.badges');
    const badgeIds = user.progress?.badges || [];

    const badges = await Badge.find({ id: { $in: badgeIds }, isActive: true })
      .sort({ order: 1, createdAt: 1 })
      .lean();

    return badges;
  } catch (error) {
    logger.error('Get user badges error:', error);
    return [];
  }
};

/**
 * Get all available badges
 */
export const getAllBadges = async (gradeLevel = 'all', category = null) => {
  try {
    const query = { isActive: true };
    
    if (gradeLevel !== 'all') {
      query.$or = [
        { gradeLevel: 'all' },
        { gradeLevel: gradeLevel }
      ];
    }

    if (category) {
      query.category = category;
    }

    const badges = await Badge.find(query)
      .sort({ order: 1, category: 1 })
      .lean();

    return badges;
  } catch (error) {
    logger.error('Get all badges error:', error);
    return [];
  }
};

/**
 * Get badge history for a user
 * Phase 3.3.3: Get award history with pagination
 */
export const getBadgeHistory = async (userId, options = {}) => {
  try {
    const {
      page = 1,
      limit = 20,
      badgeId = null
    } = options;

    const query = { userId };
    if (badgeId) {
      query.badgeId = badgeId;
    }

    const skip = (page - 1) * limit;

    const history = await BadgeHistory.find(query)
      .sort({ awardedAt: -1 })
      .skip(skip)
      .limit(limit)
      .lean();

    // Populate badge details
    const badgeIds = [...new Set(history.map(h => h.badgeId))];
    const badges = await Badge.find({ id: { $in: badgeIds } }).lean();
    const badgeMap = new Map(badges.map(b => [b.id, b]));

    const enrichedHistory = history.map(h => ({
      ...h,
      badge: badgeMap.get(h.badgeId) || null
    }));

    const total = await BadgeHistory.countDocuments(query);

    return {
      history: enrichedHistory,
      pagination: {
        page,
        limit,
        total,
        pages: Math.ceil(total / limit)
      }
    };
  } catch (error) {
    logger.error('Get badge history error:', error);
    throw error;
  }
};

File: mobile/lib/features/badges/services/badge_service.dart
/// Phase 3.3.3: Badge Service
/// Handles API calls for badge system

import '../../../core/services/api_service.dart';
import '../../../core/constants/api_endpoints.dart';
import '../models/badge_model.dart';

class BadgeService {
  final ApiService _apiService;

  BadgeService({required ApiService apiService}) : _apiService = apiService;

  /// Get all available badges
  /// [category] - Optional filter by category
  /// [gradeLevel] - Optional filter by grade level
  Future<List<Badge>> getAllBadges({
    String? category,
    String? gradeLevel,
  }) async {
    try {
      final queryParams = <String, dynamic>{};
      if (category != null) queryParams['category'] = category;
      if (gradeLevel != null) queryParams['gradeLevel'] = gradeLevel;

      final response = await _apiService.get(
        ApiEndpoints.badges,
        queryParameters: queryParams.isEmpty ? null : queryParams,
      );

      if (response.statusCode == 200 &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        final data = response.data['data'] ?? response.data;
        final badgesList = data['badges'] as List<dynamic>? ?? [];
        return badgesList
            .map((json) => Badge.fromJson(json as Map<String, dynamic>))
            .toList();
      }

      throw Exception('Failed to fetch badges');
    } catch (e) {
      print('Error fetching badges: $e');
      rethrow;
    }
  }

  /// Get specific badge by ID
  Future<Badge> getBadgeById(String badgeId) async {
    try {
      final response = await _apiService.get(
        ApiEndpoints.badge(badgeId),
      );

      if (response.statusCode == 200 &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        final data = response.data['data'] ?? response.data;
        final badgeData = data['badge'] ?? data;
        return Badge.fromJson(badgeData as Map<String, dynamic>);
      }

      throw Exception('Failed to fetch badge');
    } catch (e) {
      print('Error fetching badge: $e');
      rethrow;
    }
  }

  /// Get current user's earned badges
  Future<List<Badge>> getMyBadges() async {
    try {
      final response = await _apiService.get(ApiEndpoints.myBadges);

      if (response.statusCode == 200 &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        final data = response.data['data'] ?? response.data;
        final badgesList = data['badges'] as List<dynamic>? ?? [];
        return badgesList
            .map((json) => Badge.fromJson(json as Map<String, dynamic>))
            .toList();
      }

      throw Exception('Failed to fetch user badges');
    } catch (e) {
      print('Error fetching user badges: $e');
      rethrow;
    }
  }

  /// Get badge award history for current user
  /// [page] - Page number for pagination
  /// [limit] - Number of items per page
  /// [badgeId] - Optional filter by badge ID
  Future<Map<String, dynamic>> getBadgeHistory({
    int page = 1,
    int limit = 20,
    String? badgeId,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'page': page,
        'limit': limit,
      };
      if (badgeId != null) queryParams['badgeId'] = badgeId;

      final response = await _apiService.get(
        ApiEndpoints.badgeHistory,
        queryParameters: queryParams,
      );

      if (response.statusCode == 200 &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        final data = response.data['data'] ?? response.data;
        final historyList = data['data'] as List<dynamic>? ?? [];
        final pagination = response.data['pagination'] as Map<String, dynamic>?;

        final history = historyList
            .map((json) =>
                BadgeHistory.fromJson(json as Map<String, dynamic>))
            .toList();

        return {
          'history': history,
          'pagination': pagination ?? {},
        };
      }

      throw Exception('Failed to fetch badge history');
    } catch (e) {
      print('Error fetching badge history: $e');
      rethrow;
    }
  }

  /// Manually award badge (admin/teacher only)
  /// [badgeId] - Badge ID to award
  /// [userId] - Optional user ID (defaults to current user)
  Future<Map<String, dynamic>> awardBadge({
    required String badgeId,
    String? userId,
  }) async {
    try {
      final requestData = <String, dynamic>{};
      if (userId != null) requestData['userId'] = userId;

      final response = await _apiService.post(
        ApiEndpoints.awardBadge(badgeId),
        data: requestData.isEmpty ? null : requestData,
      );

      if ((response.statusCode == 200 || response.statusCode == 201) &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        return (response.data['data'] ?? response.data) as Map<String, dynamic>;
      }

      throw Exception('Failed to award badge');
    } catch (e) {
      print('Error awarding badge: $e');
      rethrow;
    }
  }

  /// Check and award badges for current user
  /// [triggerType] - Type of action that triggered check
  /// [triggerData] - Data related to the trigger
  Future<List<String>> checkBadges({
    String? triggerType,
    Map<String, dynamic>? triggerData,
  }) async {
    try {
      final requestData = <String, dynamic>{};
      if (triggerType != null) requestData['triggerType'] = triggerType;
      if (triggerData != null) requestData['triggerData'] = triggerData;

      final response = await _apiService.post(
        ApiEndpoints.checkBadges,
        data: requestData.isEmpty ? null : requestData,
      );

      if (response.statusCode == 200 &&
          (response.data['success'] == true ||
              response.data['success'] == 'true')) {
        final data = response.data['data'] ?? response.data;
        final newlyAwarded = data['newlyAwarded'] as List<dynamic>? ?? [];
        return newlyAwarded.map((e) => e.toString()).toList();
      }

      throw Exception('Failed to check badges');
    } catch (e) {
      print('Error checking badges: $e');
      rethrow;
    }
  }
}

